File: src/server/db/repositories.ts
import type {
  AttemptStatus,
  ConnectionInfo,
  ConnectionState,
  Draft,
  MediaItem,
  Network,
  NetworkOverride,
  PublishAttempt,
  R2SettingsView,
} from "@/lib/types";
import { type DB, getDb, newId, now } from "./index";

/* ---------------------------------- drafts --------------------------------- */

interface DraftRow {
  id: string;
  text: string;
  media_ids: string;
  networks: string;
  overrides: string;
  created_at: string;
  updated_at: string;
}

function draftFromRow(r: DraftRow): Draft {
  return {
    id: r.id,
    text: r.text,
    mediaIds: JSON.parse(r.media_ids) as string[],
    networks: JSON.parse(r.networks) as Network[],
    overrides: JSON.parse(r.overrides) as Partial<Record<Network, NetworkOverride>>,
    createdAt: r.created_at,
    updatedAt: r.updated_at,
  };
}

export const draftsRepo = (db: DB = getDb()) => ({
  create(input: Omit<Draft, "id" | "createdAt" | "updatedAt">): Draft {
    const ts = now();
    const id = newId();
    db.prepare(
      `INSERT INTO drafts (id, text, media_ids, networks, overrides, created_at, updated_at)
       VALUES (?, ?, ?, ?, ?, ?, ?)`,
    ).run(
      id,
      input.text,
      JSON.stringify(input.mediaIds),
      JSON.stringify(input.networks),
      JSON.stringify(input.overrides ?? {}),
      ts,
      ts,
    );
    return { ...input, id, createdAt: ts, updatedAt: ts };
  },

  update(
    id: string,
    input: Partial<Pick<Draft, "text" | "mediaIds" | "networks" | "overrides">>,
  ): Draft | null {
    const existing = this.get(id);
    if (!existing) return null;
    const merged = { ...existing, ...input, updatedAt: now() };
    db.prepare(
      `UPDATE drafts SET text = ?, media_ids = ?, networks = ?, overrides = ?, updated_at = ? WHERE id = ?`,
    ).run(
      merged.text,
      JSON.stringify(merged.mediaIds),
      JSON.stringify(merged.networks),
      JSON.stringify(merged.overrides),
      merged.updatedAt,
      id,
    );
    return merged;
  },

  get(id: string): Draft | null {
    const row = db.prepare("SELECT * FROM drafts WHERE id = ?").get(id) as DraftRow | undefined;
    return row ? draftFromRow(row) : null;
  },

  list(): Draft[] {
    const rows = db.prepare("SELECT * FROM drafts ORDER BY updated_at DESC").all() as DraftRow[];
    return rows.map(draftFromRow);
  },

  remove(id: string): void {
    db.prepare("DELETE FROM drafts WHERE id = ?").run(id);
  },
});

/* ------------------------------ publish attempts --------------------------- */

interface AttemptRow {
  id: string;
  draft_id: string;
  network: string;
  status: AttemptStatus;
  provider_post_id: string | null;
  provider_post_url: string | null;
  error: string | null;
  text_snapshot: string;
  created_at: string;
  updated_at: string;
}

function attemptFromRow(r: AttemptRow): PublishAttempt {
  return {
    id: r.id,
    draftId: r.draft_id,
    network: r.network as Network,
    status: r.status,
    providerPostId: r.provider_post_id ?? undefined,
    providerPostUrl: r.provider_post_url ?? undefined,
    error: r.error ?? undefined,
    textSnapshot: r.text_snapshot,
    createdAt: r.created_at,
    updatedAt: r.updated_at,
  };
}

export const attemptsRepo = (db: DB = getDb()) => ({
  create(input: { draftId: string; network: Network; textSnapshot?: string }): PublishAttempt {
    const ts = now();
    const id = newId();
    db.prepare(
      `INSERT INTO publish_attempts
         (id, draft_id, network, status, text_snapshot, created_at, updated_at)
       VALUES (?, ?, ?, 'pending', ?, ?, ?)`,
    ).run(id, input.draftId, input.network, input.textSnapshot ?? "", ts, ts);
    return this.get(id) as PublishAttempt;
  },

  setStatus(
    id: string,
    status: AttemptStatus,
    extra: { providerPostId?: string; providerPostUrl?: string; error?: string } = {},
  ): void {
    db.prepare(
      `UPDATE publish_attempts
         SET status = ?, provider_post_id = COALESCE(?, provider_post_id),
             provider_post_url = COALESCE(?, provider_post_url),
             error = ?, updated_at = ?
       WHERE id = ?`,
    ).run(
      status,
      extra.providerPostId ?? null,
      extra.providerPostUrl ?? null,
      extra.error ?? null,
      now(),
      id,
    );
  },

  get(id: string): PublishAttempt | null {
    const row = db.prepare("SELECT * FROM publish_attempts WHERE id = ?").get(id) as
      | AttemptRow
      | undefined;
    return row ? attemptFromRow(row) : null;
  },

  list(limit = 200): PublishAttempt[] {
    const rows = db
      .prepare("SELECT * FROM publish_attempts ORDER BY created_at DESC LIMIT ?")
      .all(limit) as AttemptRow[];
    return rows.map(attemptFromRow);
  },

  forDraft(draftId: string): PublishAttempt[] {
    const rows = db
      .prepare("SELECT * FROM publish_attempts WHERE draft_id = ? ORDER BY created_at DESC")
      .all(draftId) as AttemptRow[];
    return rows.map(attemptFromRow);
  },

  /** Attempts interrupted mid-flight by a process exit. */
  stuckPublishing(): PublishAttempt[] {
    const rows = db
      .prepare("SELECT * FROM publish_attempts WHERE status = 'publishing'")
      .all() as AttemptRow[];
    return rows.map(attemptFromRow);
  },
});

/* -------------------------------- connections ------------------------------ */

interface ConnectionRow {
  network: string;
  state: ConnectionState;
  credential_ref: string | null;
  account_label: string | null;
  meta: string;
  error: string | null;
  updated_at: string;
}

export const connectionsRepo = (db: DB = getDb()) => ({
  upsert(input: {
    network: Network;
    state: ConnectionState;
    credentialRef?: string | null;
    accountLabel?: string | null;
    meta?: Record<string, string>;
    error?: string | null;
  }): void {
    db.prepare(
      `INSERT INTO connections (network, state, credential_ref, account_label, meta, error, updated_at)
       VALUES (@network, @state, @credential_ref, @account_label, @meta, @error, @updated_at)
       ON CONFLICT(network) DO UPDATE SET
         state = @state,
         credential_ref = COALESCE(@credential_ref, credential_ref),
         account_label = COALESCE(@account_label, account_label),
         meta = @meta,
         error = @error,
         updated_at = @updated_at`,
    ).run({
      network: input.network,
      state: input.state,
      credential_ref: input.credentialRef ?? null,
      account_label: input.accountLabel ?? null,
      meta: JSON.stringify(input.meta ?? {}),
      error: input.error ?? null,
      updated_at: now(),
    });
  },

  get(network: Network): ConnectionInfo & { credentialRef?: string } {
    const row = db.prepare("SELECT * FROM connections WHERE network = ?").get(network) as
      | ConnectionRow
      | undefined;
    if (!row) return { network, state: "disconnected" };
    return {
      network: row.network as Network,
      state: row.state,
      accountLabel: row.account_label ?? undefined,
      meta: JSON.parse(row.meta) as Record<string, string>,
      error: row.error ?? undefined,
      credentialRef: row.credential_ref ?? undefined,
    };
  },

  list(): ConnectionInfo[] {
    const rows = db.prepare("SELECT * FROM connections").all() as ConnectionRow[];
    return rows.map((r) => ({
      network: r.network as Network,
      state: r.state,
      accountLabel: r.account_label ?? undefined,
      meta: JSON.parse(r.meta) as Record<string, string>,
      error: r.error ?? undefined,
    }));
  },

  clear(network: Network): void {
    db.prepare("DELETE FROM connections WHERE network = ?").run(network);
  },
});

/* ---------------------------------- media ---------------------------------- */

interface MediaRow {
  id: string;
  kind: "image" | "video";
  mime_type: string;
  name: string;
  size: number;
  width: number | null;
  height: number | null;
  duration_seconds: number | null;
  path: string;
  created_at: string;
}

function mediaFromRow(r: MediaRow): MediaItem {
  return {
    id: r.id,
    kind: r.kind,
    mimeType: r.mime_type,
    name: r.name,
    size: r.size,
    width: r.width ?? undefined,
    height: r.height ?? undefined,
    durationSeconds: r.duration_seconds ?? undefined,
    createdAt: r.created_at,
  };
}

export const mediaRepo = (db: DB = getDb()) => ({
  create(input: Omit<MediaItem, "id" | "createdAt"> & { path: string }): MediaItem {
    const id = newId();
    db.prepare(
      `INSERT INTO media (id, kind, mime_type, name, size, width, height, duration_seconds, path, created_at)
       VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
    ).run(
      id,
      input.kind,
      input.mimeType,
      input.name,
      input.size,
      input.width ?? null,
      input.height ?? null,
      input.durationSeconds ?? null,
      input.path,
      now(),
    );
    return this.get(id) as MediaItem;
  },

  get(id: string): MediaItem | null {
    const row = db.prepare("SELECT * FROM media WHERE id = ?").get(id) as MediaRow | undefined;
    return row ? mediaFromRow(row) : null;
  },

  /** Internal: absolute path on disk. Never expose to the client. */
  pathOf(id: string): string | null {
    const row = db.prepare("SELECT path FROM media WHERE id = ?").get(id) as
      | { path: string }
      | undefined;
    return row?.path ?? null;
  },

  remove(id: string): string | null {
    const path = this.pathOf(id);
    db.prepare("DELETE FROM media WHERE id = ?").run(id);
    return path;
  },

  byIds(ids: string[]): MediaItem[] {
    if (ids.length === 0) return [];
    const placeholders = ids.map(() => "?").join(",");
    const rows = db
      .prepare(`SELECT * FROM media WHERE id IN (${placeholders})`)
      .all(...ids) as MediaRow[];
    const map = new Map(rows.map((r) => [r.id, mediaFromRow(r)]));
    return ids.map((id) => map.get(id)).filter((m): m is MediaItem => Boolean(m));
  },
});

/* -------------------------------- oauth states ----------------------------- */

export const oauthStatesRepo = (db: DB = getDb()) => ({
  create(provider: string, state: string, verifier: string | null, ttlMs = 10 * 60 * 1000): void {
    const createdAt = now();
    const expiresAt = new Date(Date.now() + ttlMs).toISOString();
    db.prepare(
      "INSERT INTO oauth_states (state, provider, verifier, created_at, expires_at) VALUES (?, ?, ?, ?, ?)",
    ).run(state, provider, verifier, createdAt, expiresAt);
  },

  /** Single-use: returns the record and deletes it. */
  consume(provider: string, state: string): { verifier: string | null } | null {
    const row = db
      .prepare("SELECT verifier, expires_at FROM oauth_states WHERE state = ? AND provider = ?")
      .get(state, provider) as { verifier: string | null; expires_at: string } | undefined;
    db.prepare("DELETE FROM oauth_states WHERE state = ?").run(state);
    if (!row) return null;
    if (new Date(row.expires_at).getTime() < Date.now()) return null;
    return { verifier: row.verifier };
  },
});

/* -------------------------------- R2 settings ------------------------------ */

interface R2Row {
  account_id: string;
  bucket: string;
  public_url_strategy: "public" | "presigned";
  public_base_url: string | null;
  credential_ref: string | null;
}

export interface StoredR2Settings {
  accountId: string;
  bucket: string;
  publicUrlStrategy: "public" | "presigned";
  publicBaseUrl?: string;
  credentialRef?: string;
}

export const r2Repo = (db: DB = getDb()) => ({
  get(): StoredR2Settings | null {
    const row = db.prepare("SELECT * FROM r2_settings WHERE id = 1").get() as R2Row | undefined;
    if (!row) return null;
    return {
      accountId: row.account_id,
      bucket: row.bucket,
      publicUrlStrategy: row.public_url_strategy,
      publicBaseUrl: row.public_base_url ?? undefined,
      credentialRef: row.credential_ref ?? undefined,
    };
  },

  view(): R2SettingsView {
    const r2 = this.get();
    if (!r2) return { configured: false, hasCredentials: false };
    return {
      configured: Boolean(
        r2.accountId &&
          r2.credentialRef &&
          (r2.publicUrlStrategy === "presigned" || r2.publicBaseUrl),
      ),
      accountId: r2.accountId || undefined,
      bucket: r2.bucket,
      publicUrlStrategy: r2.publicUrlStrategy,
      publicBaseUrl: r2.publicBaseUrl,
      hasCredentials: Boolean(r2.credentialRef),
    };
  },

  save(input: StoredR2Settings): void {
    db.prepare(
      `INSERT INTO r2_settings (id, account_id, bucket, public_url_strategy, public_base_url, credential_ref, updated_at)
       VALUES (1, @account_id, @bucket, @public_url_strategy, @public_base_url, @credential_ref, @updated_at)
       ON CONFLICT(id) DO UPDATE SET
         account_id = @account_id, bucket = @bucket,
         public_url_strategy = @public_url_strategy, public_base_url = @public_base_url,
         credential_ref = COALESCE(@credential_ref, credential_ref), updated_at = @updated_at`,
    ).run({
      account_id: input.accountId,
      bucket: input.bucket,
      public_url_strategy: input.publicUrlStrategy,
      public_base_url: input.publicBaseUrl ?? null,
      credential_ref: input.credentialRef ?? null,
      updated_at: now(),
    });
  },

  clear(): void {
    db.prepare("DELETE FROM r2_settings WHERE id = 1").run();
  },
});

/* ------------------------------ staged objects ----------------------------- */

export const stagedRepo = (db: DB = getDb()) => ({
  add(attemptId: string | null, bucket: string, key: string): string {
    const id = newId();
    db.prepare(
      "INSERT INTO staged_objects (id, attempt_id, bucket, key, created_at) VALUES (?, ?, ?, ?, ?)",
    ).run(id, attemptId, bucket, key, now());
    return id;
  },

  remove(id: string): void {
    db.prepare("DELETE FROM staged_objects WHERE id = ?").run(id);
  },

  forAttempt(attemptId: string): { id: string; bucket: string; key: string }[] {
    return db
      .prepare("SELECT id, bucket, key FROM staged_objects WHERE attempt_id = ?")
      .all(attemptId) as { id: string; bucket: string; key: string }[];
  },

  olderThan(cutoffIso: string): { id: string; bucket: string; key: string }[] {
    return db
      .prepare("SELECT id, bucket, key FROM staged_objects WHERE created_at < ?")
      .all(cutoffIso) as { id: string; bucket: string; key: string }[];
  },
});

File: Blather/Persistence/Repositories.swift
import Foundation
import GRDB

struct DraftsRepository: Sendable {
    let dbQueue: DatabaseQueue

    func create(
        text: String,
        mediaIds: [String],
        accountIds: [String],
        networks: [Network],
        overrides: [Network: NetworkOverride]
    ) throws -> Draft {
        let ts = Time.now()
        let id = Time.newId()
        try dbQueue.write { db in
            try db.execute(
                sql: """
                INSERT INTO drafts (id, text, media_ids, account_ids, networks, overrides, created_at, updated_at)
                VALUES (?, ?, ?, ?, ?, ?, ?, ?)
                """,
                arguments: [
                    id,
                    text,
                    JSONCodec.encode(mediaIds),
                    JSONCodec.encode(accountIds),
                    JSONCodec.encode(networks.map(\.rawValue)),
                    JSONCodec.encodeOverrides(overrides),
                    ts,
                    ts,
                ]
            )
        }
        return Draft(
            id: id,
            text: text,
            mediaIds: mediaIds,
            accountIds: accountIds,
            networks: networks,
            overrides: overrides,
            createdAt: ts,
            updatedAt: ts
        )
    }

    func update(
        id: String,
        text: String? = nil,
        mediaIds: [String]? = nil,
        accountIds: [String]? = nil,
        networks: [Network]? = nil,
        overrides: [Network: NetworkOverride]? = nil
    ) throws -> Draft? {
        guard var existing = try get(id) else { return nil }
        if let text { existing.text = text }
        if let mediaIds { existing.mediaIds = mediaIds }
        if let accountIds { existing.accountIds = accountIds }
        if let networks { existing.networks = networks }
        if let overrides { existing.overrides = overrides }
        existing.updatedAt = Time.now()
        try dbQueue.write { db in
            try db.execute(
                sql: """
                UPDATE drafts
                SET text = ?, media_ids = ?, account_ids = ?, networks = ?, overrides = ?, updated_at = ?
                WHERE id = ?
                """,
                arguments: [
                    existing.text,
                    JSONCodec.encode(existing.mediaIds),
                    JSONCodec.encode(existing.accountIds),
                    JSONCodec.encode(existing.networks.map(\.rawValue)),
                    JSONCodec.encodeOverrides(existing.overrides),
                    existing.updatedAt,
                    id,
                ]
            )
        }
        return existing
    }

    func get(_ id: String) throws -> Draft? {
        try dbQueue.read { db in
            try Row.fetchOne(db, sql: "SELECT * FROM drafts WHERE id = ?", arguments: [id]).map(Self.draft(from:))
        }
    }

    func list() throws -> [Draft] {
        try dbQueue.read { db in
            try Row.fetchAll(db, sql: "SELECT * FROM drafts ORDER BY updated_at DESC").map(Self.draft(from:))
        }
    }

    func remove(_ id: String) throws {
        try dbQueue.write { db in
            try db.execute(sql: "DELETE FROM drafts WHERE id = ?", arguments: [id])
        }
    }

    private static func draft(from row: Row) -> Draft {
        let networkIds: [String] = JSONCodec.decode([String].self, from: row["networks"])
        return Draft(
            id: row["id"],
            text: row["text"],
            mediaIds: JSONCodec.decode([String].self, from: row["media_ids"]),
            accountIds: JSONCodec.decode([String].self, from: row["account_ids"]),
            networks: networkIds.compactMap(Network.init(rawValue:)),
            overrides: JSONCodec.decodeOverrides(row["overrides"]),
            createdAt: row["created_at"],
            updatedAt: row["updated_at"]
        )
    }
}

struct AttemptsRepository: Sendable {
    let dbQueue: DatabaseQueue

    func create(
        draftId: String,
        network: Network,
        accountId: String,
        accountLabelSnapshot: String? = nil,
        textSnapshot: String = ""
    ) throws -> PublishAttempt {
        let ts = Time.now()
        let id = Time.newId()
        try dbQueue.write { db in
            try db.execute(
                sql: """
                INSERT INTO publish_attempts
                  (id, draft_id, network, account_id, account_label_snapshot, status, text_snapshot, created_at, updated_at)
                VALUES (?, ?, ?, ?, ?, 'pending', ?, ?, ?)
                """,
                arguments: [
                    id,
                    draftId,
                    network.rawValue,
                    accountId,
                    accountLabelSnapshot,
                    textSnapshot,
                    ts,
                    ts,
                ]
            )
        }
        return try get(id)!
    }

    func setStatus(
        id: String,
        status: AttemptStatus,
        providerPostId: String? = nil,
        providerPostUrl: String? = nil,
        error: String? = nil
    ) throws {
        try dbQueue.write { db in
            try db.execute(
                sql: """
                UPDATE publish_attempts
                   SET status = ?,
                       provider_post_id = COALESCE(?, provider_post_id),
                       provider_post_url = COALESCE(?, provider_post_url),
                       error = ?,
                       updated_at = ?
                 WHERE id = ?
                """,
                arguments: [status.rawValue, providerPostId, providerPostUrl, error, Time.now(), id]
            )
        }
    }

    func get(_ id: String) throws -> PublishAttempt? {
        try dbQueue.read { db in
            try Row.fetchOne(db, sql: "SELECT * FROM publish_attempts WHERE id = ?", arguments: [id])
                .map(Self.attempt(from:))
        }
    }

    func list(limit: Int = 200) throws -> [PublishAttempt] {
        try dbQueue.read { db in
            try Row.fetchAll(
                db,
                sql: "SELECT * FROM publish_attempts ORDER BY created_at DESC LIMIT ?",
                arguments: [limit]
            ).map(Self.attempt(from:))
        }
    }

    func forDraft(_ draftId: String) throws -> [PublishAttempt] {
        try dbQueue.read { db in
            try Row.fetchAll(
                db,
                sql: "SELECT * FROM publish_attempts WHERE draft_id = ? ORDER BY created_at DESC",
                arguments: [draftId]
            ).map(Self.attempt(from:))
        }
    }

    func stuckPublishing() throws -> [PublishAttempt] {
        try dbQueue.read { db in
            try Row.fetchAll(db, sql: "SELECT * FROM publish_attempts WHERE status = 'publishing'")
                .map(Self.attempt(from:))
        }
    }

    private static func attempt(from row: Row) -> PublishAttempt {
        PublishAttempt(
            id: row["id"],
            draftId: row["draft_id"],
            network: Network(rawValue: row["network"]) ?? .x,
            accountId: row["account_id"],
            accountLabelSnapshot: row["account_label_snapshot"],
            status: AttemptStatus(rawValue: row["status"]) ?? .pending,
            providerPostId: row["provider_post_id"],
            providerPostUrl: row["provider_post_url"],
            error: row["error"],
            textSnapshot: row["text_snapshot"],
            createdAt: row["created_at"],
            updatedAt: row["updated_at"]
        )
    }
}

struct ConnectionsRepository: Sendable {
    let dbQueue: DatabaseQueue

    func upsert(
        accountId: String,
        network: Network,
        providerAccountId: String? = nil,
        state: ConnectionState,
        credentialRef: String? = nil,
        accountLabel: String? = nil,
        meta: [String: String] = [:],
        error: String? = nil,
        isRemoved: Bool = false
    ) throws {
        try dbQueue.write { db in
            try db.execute(
                sql: """
                INSERT INTO social_accounts
                  (id, network, provider_account_id, state, credential_ref, account_label, meta, error, is_removed, updated_at)
                VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
                ON CONFLICT(id) DO UPDATE SET
                  network = excluded.network,
                  provider_account_id = COALESCE(excluded.provider_account_id, social_accounts.provider_account_id),
                  state = excluded.state,
                  credential_ref = COALESCE(excluded.credential_ref, social_accounts.credential_ref),
                  account_label = COALESCE(excluded.account_label, social_accounts.account_label),
                  meta = excluded.meta,
                  error = excluded.error,
                  is_removed = excluded.is_removed,
                  updated_at = excluded.updated_at
                """,
                arguments: [
                    accountId,
                    network.rawValue,
                    providerAccountId,
                    state.rawValue,
                    credentialRef,
                    accountLabel,
                    JSONCodec.encode(meta),
                    error,
                    isRemoved,
                    Time.now(),
                ]
            )
        }
    }

    func get(_ accountId: String) throws -> ConnectionInfo? {
        try dbQueue.read { db in
            try Row.fetchOne(
                db,
                sql: "SELECT * FROM social_accounts WHERE id = ?",
                arguments: [accountId]
            ).map(Self.info(from:))
        }
    }

    func find(network: Network, providerAccountId: String) throws -> ConnectionInfo? {
        try dbQueue.read { db in
            try Row.fetchOne(
                db,
                sql: "SELECT * FROM social_accounts WHERE network = ? AND provider_account_id = ?",
                arguments: [network.rawValue, providerAccountId]
            ).map(Self.info(from:))
        }
    }

    func list(includeRemoved: Bool = false) throws -> [ConnectionInfo] {
        try dbQueue.read { db in
            let sql = includeRemoved
                ? "SELECT * FROM social_accounts ORDER BY network, account_label, updated_at"
                : "SELECT * FROM social_accounts WHERE is_removed = 0 ORDER BY network, account_label, updated_at"
            return try Row.fetchAll(db, sql: sql).map(Self.info(from:))
        }
    }

    func list(network: Network, includeRemoved: Bool = false) throws -> [ConnectionInfo] {
        try dbQueue.read { db in
            let removedClause = includeRemoved ? "" : "AND is_removed = 0"
            return try Row.fetchAll(
                db,
                sql: """
                SELECT * FROM social_accounts
                WHERE network = ? \(removedClause)
                ORDER BY account_label, updated_at
                """,
                arguments: [network.rawValue]
            ).map(Self.info(from:))
        }
    }

    func clear(_ accountId: String) throws {
        try dbQueue.write { db in
            try db.execute(
                sql: """
                UPDATE social_accounts
                SET state = 'disconnected', credential_ref = NULL, error = NULL,
                    is_removed = 1, updated_at = ?
                WHERE id = ?
                """,
                arguments: [Time.now(), accountId]
            )
        }
    }

    private static func info(from row: Row) -> ConnectionInfo {
        ConnectionInfo(
            id: row["id"],
            network: Network(rawValue: row["network"]) ?? .x,
            providerAccountId: row["provider_account_id"],
            state: ConnectionState(rawValue: row["state"]) ?? .disconnected,
            accountLabel: row["account_label"],
            meta: JSONCodec.decode([String: String].self, from: row["meta"]),
            error: row["error"],
            credentialRef: row["credential_ref"],
            isRemoved: row["is_removed"]
        )
    }
}

struct MediaRepository: Sendable {
    let dbQueue: DatabaseQueue

    func create(
        kind: MediaKind,
        mimeType: String,
        name: String,
        size: Int,
        width: Int? = nil,
        height: Int? = nil,
        durationSeconds: Double? = nil,
        path: String
    ) throws -> MediaItem {
        let id = Time.newId()
        let ts = Time.now()
        try dbQueue.write { db in
            try db.execute(
                sql: """
                INSERT INTO media (id, kind, mime_type, name, size, width, height, duration_seconds, path, created_at)
                VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
                """,
                arguments: [id, kind.rawValue, mimeType, name, size, width, height, durationSeconds, path, ts]
            )
        }
        return MediaItem(
            id: id,
            kind: kind,
            mimeType: mimeType,
            name: name,
            size: size,
            width: width,
            height: height,
            durationSeconds: durationSeconds,
            createdAt: ts
        )
    }

    func get(_ id: String) throws -> MediaItem? {
        try dbQueue.read { db in
            try Row.fetchOne(db, sql: "SELECT * FROM media WHERE id = ?", arguments: [id]).map(Self.item(from:))
        }
    }

    func pathOf(_ id: String) throws -> String? {
        try dbQueue.read { db in
            try String.fetchOne(db, sql: "SELECT path FROM media WHERE id = ?", arguments: [id])
        }
    }

    func remove(_ id: String) throws -> String? {
        let path = try pathOf(id)
        try dbQueue.write { db in
            try db.execute(sql: "DELETE FROM media WHERE id = ?", arguments: [id])
        }
        return path
    }

    func byIds(_ ids: [String]) throws -> [MediaItem] {
        guard !ids.isEmpty else { return [] }
        let placeholders = Array(repeating: "?", count: ids.count).joined(separator: ",")
        let rows = try dbQueue.read { db in
            try Row.fetchAll(db, sql: "SELECT * FROM media WHERE id IN (\(placeholders))", arguments: StatementArguments(ids))
        }
        let map = Dictionary(uniqueKeysWithValues: rows.map { ($0["id"] as String, Self.item(from: $0)) })
        return ids.compactMap { map[$0] }
    }

    private static func item(from row: Row) -> MediaItem {
        MediaItem(
            id: row["id"],
            kind: MediaKind(rawValue: row["kind"]) ?? .image,
            mimeType: row["mime_type"],
            name: row["name"],
            size: row["size"],
            width: row["width"],
            height: row["height"],
            durationSeconds: row["duration_seconds"],
            createdAt: row["created_at"]
        )
    }
}

struct OAuthStatesRepository: Sendable {
    let dbQueue: DatabaseQueue

    func create(provider: String, state: String, verifier: String?, ttlMs: Int = 10 * 60 * 1000) throws {
        let createdAt = Time.now()
        let expiresAt = Time.isoFormatter.string(from: Date().addingTimeInterval(Double(ttlMs) / 1000))
        try dbQueue.write { db in
            try db.execute(
                sql: "INSERT INTO oauth_states (state, provider, verifier, created_at, expires_at) VALUES (?, ?, ?, ?, ?)",
                arguments: [state, provider, verifier, createdAt, expiresAt]
            )
        }
    }

    func consume(provider: String, state: String) throws -> String? {
        try dbQueue.write { db in
            let row = try Row.fetchOne(
                db,
                sql: "SELECT verifier, expires_at FROM oauth_states WHERE state = ? AND provider = ?",
                arguments: [state, provider]
            )
            try db.execute(sql: "DELETE FROM oauth_states WHERE state = ?", arguments: [state])
            guard let row else { return nil }
            let expiresAt: String = row["expires_at"]
            if let date = Time.isoFormatter.date(from: expiresAt), date < Date() {
                return nil
            }
            return row["verifier"]
        }
    }
}

struct StoredR2Settings: Hashable, Sendable {
    var accountId: String
    var bucket: String
    var publicUrlStrategy: String
    var publicBaseUrl: String?
    var credentialRef: String?
    var jurisdiction: String? = nil
}

struct R2SettingsRepository: Sendable {
    let dbQueue: DatabaseQueue

    func get() throws -> StoredR2Settings? {
        try dbQueue.read { db in
            guard let row = try Row.fetchOne(db, sql: "SELECT * FROM r2_settings WHERE id = 1") else { return nil }
            return StoredR2Settings(
                accountId: row["account_id"],
                bucket: row["bucket"],
                publicUrlStrategy: row["public_url_strategy"],
                publicBaseUrl: row["public_base_url"],
                credentialRef: row["credential_ref"],
                jurisdiction: row["jurisdiction"]
            )
        }
    }

    func view() throws -> R2SettingsView {
        guard let r2 = try get() else {
            return R2SettingsView(configured: false, hasCredentials: false)
        }
        let configured = !r2.accountId.isEmpty
            && r2.credentialRef != nil
            && (r2.publicUrlStrategy == "presigned" || r2.publicBaseUrl != nil)
        return R2SettingsView(
            configured: configured,
            accountId: r2.accountId.isEmpty ? nil : r2.accountId,
            bucket: r2.bucket,
            publicUrlStrategy: r2.publicUrlStrategy,
            publicBaseUrl: r2.publicBaseUrl,
            hasCredentials: r2.credentialRef != nil,
            jurisdiction: r2.jurisdiction
        )
    }

    func save(_ input: StoredR2Settings) throws {
        try dbQueue.write { db in
            try db.execute(
                sql: """
                INSERT INTO r2_settings (id, account_id, bucket, public_url_strategy, public_base_url, credential_ref, jurisdiction, updated_at)
                VALUES (1, ?, ?, ?, ?, ?, ?, ?)
                ON CONFLICT(id) DO UPDATE SET
                  account_id = excluded.account_id,
                  bucket = excluded.bucket,
                  public_url_strategy = excluded.public_url_strategy,
                  public_base_url = excluded.public_base_url,
                  credential_ref = COALESCE(excluded.credential_ref, r2_settings.credential_ref),
                  jurisdiction = excluded.jurisdiction,
                  updated_at = excluded.updated_at
                """,
                arguments: [
                    input.accountId,
                    input.bucket,
                    input.publicUrlStrategy,
                    input.publicBaseUrl,
                    input.credentialRef,
                    input.jurisdiction,
                    Time.now(),
                ]
            )
        }
    }

    func clear() throws {
        try dbQueue.write { db in
            try db.execute(sql: "DELETE FROM r2_settings WHERE id = 1")
        }
    }
}

struct StagedRepository: Sendable {
    let dbQueue: DatabaseQueue

    func add(attemptId: String?, bucket: String, key: String) throws -> String {
        let id = Time.newId()
        try dbQueue.write { db in
            try db.execute(
                sql: "INSERT INTO staged_objects (id, attempt_id, bucket, key, created_at) VALUES (?, ?, ?, ?, ?)",
                arguments: [id, attemptId, bucket, key, Time.now()]
            )
        }
        return id
    }

    func remove(_ id: String) throws {
        try dbQueue.write { db in
            try db.execute(sql: "DELETE FROM staged_objects WHERE id = ?", arguments: [id])
        }
    }

    func forAttempt(_ attemptId: String) throws -> [(id: String, bucket: String, key: String)] {
        try dbQueue.read { db in
            try Row.fetchAll(
                db,
                sql: "SELECT id, bucket, key FROM staged_objects WHERE attempt_id = ?",
                arguments: [attemptId]
            ).map { (id: $0["id"], bucket: $0["bucket"], key: $0["key"]) }
        }
    }

    func olderThan(_ cutoffIso: String) throws -> [(id: String, bucket: String, key: String)] {
        try dbQueue.read { db in
            try Row.fetchAll(
                db,
                sql: "SELECT id, bucket, key FROM staged_objects WHERE created_at < ?",
                arguments: [cutoffIso]
            ).map { (id: $0["id"], bucket: $0["bucket"], key: $0["key"]) }
        }
    }
}

extension AppDatabase {
    var drafts: DraftsRepository { DraftsRepository(dbQueue: dbQueue) }
    var attempts: AttemptsRepository { AttemptsRepository(dbQueue: dbQueue) }
    var connections: ConnectionsRepository { ConnectionsRepository(dbQueue: dbQueue) }
    var media: MediaRepository { MediaRepository(dbQueue: dbQueue) }
    var oauthStates: OAuthStatesRepository { OAuthStatesRepository(dbQueue: dbQueue) }
    var r2: R2SettingsRepository { R2SettingsRepository(dbQueue: dbQueue) }
    var staged: StagedRepository { StagedRepository(dbQueue: dbQueue) }
}
